builtin/providers/softlayer/resource_softlayer_user_customer.go

```diff
@@ -0,0 +1,400 @@
+		userObj.Country = d.Get("country").(string)
+	}
+	if d.HasChange("timezone") {
+		userObj.Timezone = d.Get("timezone").(int)
+	}
+	if d.HasChange("user_status") {
+		userObj.UserStatus = d.Get("user_status").(int)
+	}
+
+	_, err = client.EditObject(sluid, userObj)
+	if err != nil {
+		return fmt.Errorf("Error received while editing softlayer_user: %s", err)
+	}
+
+	if d.HasChange("permissions") {
+		// TODO Use set math functions (in schema.Set) to compute the difference, vs clearing and re-adding permissions
+		old, new := d.GetChange("permissions")
+
+		oldPermissions := make([]datatypes.SoftLayer_User_Customer_CustomerPermission_Permission, 0, old.(*schema.Set).Len())
+		newPermissions := make([]datatypes.SoftLayer_User_Customer_CustomerPermission_Permission, 0, new.(*schema.Set).Len())
+
+		for _, elem := range old.(*schema.Set).List() {
+			oldPermissions = append(oldPermissions, makePermission(elem.(string)))
+		}
+
+		for _, elem := range new.(*schema.Set).List() {
+			newPermissions = append(newPermissions, makePermission(elem.(string)))
+		}
+
+		// 'remove' all old permissions
+		err = client.RemoveBulkPortalPermission(sluid, oldPermissions)
+		if err != nil {
+			return fmt.Errorf("Error received while removing old permissions from softlayer_user: %s", err)
+		}
+
+		// 'add' new permission set
+		err = client.AddBulkPortalPermission(sluid, newPermissions)
+		if err != nil {
+			return fmt.Errorf("Error received while assigning new permissions to softlayer_user: %s", err)
+		}
+	}
+
+	if d.HasChange("has_api_key") {
+		api_key_flag := d.Get("has_api_key").(bool)
+		// if true, then it means create an api key if none exists. Its a no-op if an api key already exists.
+		// else false means, delete the api key if one exists. Its a no-op if no api key exists.
+		keys, err := client.GetApiAuthenticationKeys(sluid)
+		if err != nil {
+			return fmt.Errorf("Error fetching the API key editing softlayer_user resource: %s", err)
+		}
+		// Create a key if flag is true, and a key does not already exisit.
+		if api_key_flag {
+			if len(keys) == 0 { // means key does not exist, so create one.
+				err = client.AddApiAuthenticationKey(sluid)
+				if err != nil {
+					return fmt.Errorf("Error creating API key while editing softlayer_user resource: %s", err)
+				}
+				// get the details of the new one, so that you can update the terraform resource.
+				keys, err := client.GetApiAuthenticationKeys(sluid)
+				if err != nil {
+					return fmt.Errorf("Error fetching the newly created API key while editing softlayer_user resource: %s", err)
+				}
+				d.Set("api_key", keys[0].AuthenticationKey) // as api_key is a computed field
+			}
+		} else {
+			// If false, then delete the key if there was one.
+			if len(keys) > 0 {
+				_, err = client.RemoveApiAuthenticationKey(sluid, keys)
+				d.Set("api_key", nil)
+				if err != nil {
+					return fmt.Errorf("Error deleting API key while editing softlayer_user resource: %s", err)
+				}
+			}
```

Comment:
I think when the api_key_flag is false, we may want to unset it from the computed field? regardless of whether the apis returned keys or not.